src/ceph-disk

```diff
@@ -2461,6 +2678,316 @@ def main_activate(args):
+           if mounted_path is None:
+               raise Error('%s is not mounted' % dev_path)
+           osd_id = get_osd_id(mounted_path)
+           mount_info.append(dev_path)
+           mount_info.append(mounted_path)
+
+    # Do not do anything if osd is already down.
+    status_code = _check_osd_status(args.cluster, osd_id)
+    if status_code == OSD_STATUS_IN_UP:
+        if args.mark_out is True:
+            _mark_osd_out(args.cluster, osd_id)
+        stop_daemon(args.cluster, osd_id)
+    elif status_code == OSD_STATUS_IN_DOWN:
+        if args.mark_out is True:
+            _mark_osd_out(args.cluster, osd_id)
+        LOG.info("OSD already out/down. Do not do anything now.")
+        return
+    elif status_code == OSD_STATUS_OUT_UP:
+        stop_daemon(args.cluster, osd_id)
+    elif status_code == OSD_STATUS_OUT_DOWN:
+        LOG.info("OSD already out/down. Do not do anything now.")
+        return
+
+    # GET the mounted device and mount point.
+    # If we already get mount_info (with specific parameter), pass this stage
+    if not mount_info:
+        mount_info = convert_osd_id(args.cluster, osd_id)
+
+    # remove 'ready', 'active', and INIT-specific files.
+    _remove_osd_directory_files(mount_info[1], args.cluster)
+
+    # Write deactivate to osd directory!
+    with file(os.path.join(mount_info[1], 'deactive'), 'w'):
+        path_set_context(os.path.join(mount_info[1], 'deactive'))
+        pass
+
+    unmount(mount_info[1])
+    LOG.info("Umount `%s` successfully.", mount_info[1])
+
+    return
+
+###########################
+
+def _remove_from_crush_map(cluster, osd_id):
+    LOG.info("Prepare to remove osd.%s from crush map..." % osd_id)
+    try:
+        out, ret = command(
+                [
+                    'ceph',
+                    'osd',
+                    'crush',
+                    'remove',
+                    'osd.%s' % osd_id,
+                    ],
+                )
+    except subprocess.CalledProcessError as e:
+        raise Error(e)
+
+def _delete_osd_auth_key(cluster, osd_id):
+    LOG.info("Prepare to delete osd.%s cephx key..." % osd_id)
+    try:
+        out, ret = command(
+                [
+                    'ceph',
+                    'auth',
+                    'del',
+                    'osd.%s' % osd_id,
+                    ],
+                )
+    except subprocess.CalledProcessError as e:
+        raise Error(e)
+
+def _deallocate_osd_id(cluster, osd_id):
+    LOG.info("Prepare to deallocate the osd-id: %s..." % osd_id)
+    try:
+        out, ret = command(
+                [
+                    'ceph',
+                    'osd',
+                    'rm',
+                    '%s' % osd_id,
+                    ],
+                )
+    except subprocess.CalledProcessError as e:
+        raise Error(e)
+
+def main_destroy(args):
+    dmcrypt = False
+    mount_info = []
+    # get everything we need when we start destroy.
+    if args.destroy_by_id:
```

Comment:
Sure, but the most information on `list_devices` function I did not use it.
I am afraid that will increase the redundant overhead when get something information I don't need.

that's trade-off, I don't know how to chose :p 
use list_devices makes code simplely! or we do not get redundant information?